File: backend/src/models/preferences.py
from __future__ import annotations

from datetime import datetime
from typing import Annotated, Literal

from pydantic import BaseModel, Field, field_validator

PreferenceCategory = Literal[
    "diet",
    "pace",
    "budget",
    "interest",
    "accommodation",
    "transport",
]
PreferenceSource = Literal["inferred", "manual"]


class PreferenceItem(BaseModel):
    id: str
    category: PreferenceCategory
    content: str
    source: PreferenceSource
    confidence: float = Field(ge=0, le=1)
    is_active: bool
    created_at: datetime | None = None
    updated_at: datetime | None = None
    deleted_at: datetime | None = None


class PreferenceSummary(BaseModel):
    total: int
    active_count: int
    categories: dict[str, int]


class PreferenceListData(BaseModel):
    preferences: list[PreferenceItem]
    summary: PreferenceSummary


class AddPreferenceRequest(BaseModel):
    category: PreferenceCategory
    content: str = Field(min_length=1)

    @field_validator("content")
    @classmethod
    def strip_content(cls, value: str) -> str:
        value = value.strip()
        if not value:
            raise ValueError("content must not be empty")
        return value


class UpdatePreferenceRequest(BaseModel):
    content: str = Field(min_length=1)

    @field_validator("content")
    @classmethod
    def strip_content(cls, value: str) -> str:
        value = value.strip()
        if not value:
            raise ValueError("content must not be empty")
        return value


class PreferenceDeletionData(BaseModel):
    id: str
    is_active: bool
    deleted_at: datetime


class AddPreferenceOperation(AddPreferenceRequest):
    """批量操作中的新增偏好项。"""

    action: Literal["add"] = "add"


class UpdatePreferenceOperation(UpdatePreferenceRequest):
    """批量操作中的修改偏好项。"""

    action: Literal["update"] = "update"
    id: str = Field(min_length=1)


class DeletePreferenceRequest(BaseModel):
    """批量操作中的删除偏好项。"""

    action: Literal["delete"] = "delete"
    id: str = Field(min_length=1)


PreferenceOperation = Annotated[
    AddPreferenceOperation | UpdatePreferenceOperation | DeletePreferenceRequest,
    Field(discriminator="action"),
]


class BatchPreferenceRequest(BaseModel):
    """用户画像说明书中定义的批量偏好操作请求。"""

    action: Literal["batch"] = "batch"
    operations: list[PreferenceOperation] = Field(min_length=1)


class PreferenceItemInput(BaseModel):
    """整体替换偏好时单条输入。"""

    category: PreferenceCategory
    content: str = Field(min_length=1)

    @field_validator("content")
    @classmethod
    def strip_content(cls, value: str) -> str:
        value = value.strip()
        if not value:
            raise ValueError("content must not be empty")
        return value


class ReplacePreferencesRequest(BaseModel):
    """整体替换用户手动偏好（保留推断偏好不动）。"""

    items: list[PreferenceItemInput] = Field(default_factory=list)

File: frontend/src/api/preferences.ts
import { API_BASE_URL, USER_ID } from './chat'

// 对应后端 PreferenceCategory（src/models/preferences.py）
export type PreferenceCategory =
  | 'diet'
  | 'pace'
  | 'budget'
  | 'interest'
  | 'accommodation'
  | 'transport'

export type PreferenceSource = 'inferred' | 'manual'

export type PreferenceItem = {
  id: string
  category: PreferenceCategory
  content: string
  source: PreferenceSource
  confidence: number
  is_active: boolean
  created_at?: string | null
  updated_at?: string | null
  deleted_at?: string | null
}

export type PreferenceSummary = {
  total: number
  active_count: number
  categories: Record<string, number>
}

export type PreferenceListData = {
  preferences: PreferenceItem[]
  summary: PreferenceSummary
}

type PreferenceListResponse = {
  code: number
  message: string
  data: PreferenceListData
}

export type PreferenceItemInput = {
  category: PreferenceCategory
  content: string
}

const EMPTY_DATA: PreferenceListData = {
  preferences: [],
  summary: { total: 0, active_count: 0, categories: {} },
}

/**
 * GET /api/v1/users/me/preferences
 * 拉取当前用户的全部偏好（含推断）。前端按 source=manual 重建表单。
 */
export async function fetchPreferences(signal?: AbortSignal): Promise<PreferenceListData> {
  const url = `${API_BASE_URL}/users/me/preferences?include_inferred=true`
  const response = await fetch(url, {
    method: 'GET',
    headers: { 'X-User-Id': USER_ID },
    signal,
  })

  if (!response.ok) {
    const message = await response.text()
    throw new Error(message || response.statusText || '获取偏好失败')
  }

  const payload = (await response.json()) as PreferenceListResponse
  return payload.data ?? EMPTY_DATA
}

/**
 * PUT /api/v1/users/me/preferences
 * 整体替换当前用户的手动偏好（推断偏好保留不动）。
 */
export async function replacePreferences(
  items: PreferenceItemInput[],
  signal?: AbortSignal,
): Promise<PreferenceListData> {
  const url = `${API_BASE_URL}/users/me/preferences`
  const response = await fetch(url, {
    method: 'PUT',
    headers: {
      'X-User-Id': USER_ID,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ items }),
    signal,
  })

  if (!response.ok) {
    const message = await response.text()
    throw new Error(message || response.statusText || '保存偏好失败')
  }

  const payload = (await response.json()) as PreferenceListResponse
  return payload.data ?? EMPTY_DATA
}
